File: src/content_factory/api/routers/timeline.py
"""Video studio: timeline NLE, render plan, video and voiceover."""

from __future__ import annotations

from fastapi import (
    APIRouter,
    File,
    HTTPException,
    UploadFile,
)
from fastapi.responses import FileResponse

from ...models import (
    AiAssistRequest,
    MarkerCreate,
    Project,
    RenderPlan,
    RenderRequest,
    SceneAudioRequest,
    SceneBulkRequest,
    SceneClipResponse,
    SceneMoveRequest,
    ScenePasteRequest,
    SceneReverseRequest,
    SceneSpeedRequest,
    SceneSplitRequest,
    SceneTrimRequest,
    StructuredTimeline,
    TimelineReport,
    VideoEditUpdate,
    VideoProject,
)
from ...render import RenderError
from ...service import ContentFactoryService, NotFoundError
from ..deps import get_or_404, guard, guard_value


def build_router(service: ContentFactoryService) -> APIRouter:

    router = APIRouter()

    @router.get("/projects/{project_id}/video-project", response_model=VideoProject)
    async def get_video_project(project_id: str) -> VideoProject:
        project = get_or_404(service, project_id)
        if project.video_project is None:
            raise HTTPException(
                status_code=404, detail="No video project has been built yet"
            )
        return project.video_project

    @router.put("/projects/{project_id}/video-project", response_model=Project)
    async def update_video_project(project_id: str, data: VideoEditUpdate) -> Project:
        return guard(lambda: service.update_video_project(project_id, data.project))

    @router.get("/projects/{project_id}/timeline/report", response_model=TimelineReport)
    async def timeline_report(project_id: str) -> TimelineReport:
        return guard_value(lambda: service.timeline_report(project_id))

    @router.get("/projects/{project_id}/render-plan", response_model=RenderPlan)
    async def render_plan(project_id: str) -> RenderPlan:
        return guard_value(lambda: service.render_plan(project_id))

    @router.post("/projects/{project_id}/timeline/normalize", response_model=Project)
    async def normalize_timeline(project_id: str) -> Project:
        return guard_value(lambda: service.normalize_video_project(project_id))

    @router.post(
        "/projects/{project_id}/timeline/scenes/{scene_id}/split",
        response_model=Project,
    )
    async def split_scene(
        project_id: str, scene_id: str, data: SceneSplitRequest
    ) -> Project:
        return guard_value(
            lambda: service.split_video_scene(project_id, scene_id, at=data.at)
        )

    @router.post(
        "/projects/{project_id}/timeline/scenes/{scene_id}/merge",
        response_model=Project,
    )
    async def merge_scene(project_id: str, scene_id: str) -> Project:
        return guard_value(lambda: service.merge_video_scene(project_id, scene_id))

    @router.post(
        "/projects/{project_id}/timeline/scenes/{scene_id}/duplicate",
        response_model=Project,
    )
    async def duplicate_scene(project_id: str, scene_id: str) -> Project:
        return guard_value(lambda: service.duplicate_video_scene(project_id, scene_id))

    @router.delete(
        "/projects/{project_id}/timeline/scenes/{scene_id}", response_model=Project
    )
    async def delete_scene(project_id: str, scene_id: str) -> Project:
        return guard_value(lambda: service.delete_video_scene(project_id, scene_id))

    @router.post(
        "/projects/{project_id}/timeline/scenes/{scene_id}/move",
        response_model=Project,
    )
    async def move_scene(
        project_id: str, scene_id: str, data: SceneMoveRequest
    ) -> Project:
        return guard_value(
            lambda: service.move_video_scene(project_id, scene_id, data.to_index)
        )

    @router.post("/projects/{project_id}/timeline/scenes/bulk", response_model=Project)
    async def bulk_update_scenes(project_id: str, data: SceneBulkRequest) -> Project:
        return guard_value(
            lambda: service.bulk_update_video_scenes(
                project_id, data.scene_ids, dict(data.patch)
            )
        )

    @router.post("/projects/{project_id}/timeline/markers", response_model=Project)
    async def add_marker(project_id: str, data: MarkerCreate) -> Project:
        return guard_value(
            lambda: service.add_timeline_marker(
                project_id, data.time_seconds, data.label, data.color
            )
        )

    @router.post(
        "/projects/{project_id}/timeline/scenes/{scene_id}/speed",
        response_model=Project,
    )
    async def set_scene_speed(
        project_id: str, scene_id: str, data: SceneSpeedRequest
    ) -> Project:
        """Retime a scene (0.5x slow-mo .. 2x fast-forward)."""
        return guard_value(
            lambda: service.set_scene_speed(project_id, scene_id, data.speed)
        )

    @router.post(
        "/projects/{project_id}/timeline/scenes/{scene_id}/reverse",
        response_model=Project,
    )
    async def reverse_scene(
        project_id: str, scene_id: str, data: SceneReverseRequest
    ) -> Project:
        """Play a scene's source media backwards (boomerang)."""
        return guard_value(
            lambda: service.reverse_scene(project_id, scene_id, data.reverse)
        )

    @router.post(
        "/projects/{project_id}/timeline/scenes/{scene_id}/trim",
        response_model=Project,
    )
    async def trim_scene(
        project_id: str, scene_id: str, data: SceneTrimRequest
    ) -> Project:
        """Set a scene's source in/out handles without moving the timeline."""
        return guard_value(
            lambda: service.trim_scene(
                project_id, scene_id, data.trim_start, data.trim_end
            )
        )

    @router.post(
        "/projects/{project_id}/timeline/scenes/{scene_id}/audio",
        response_model=Project,
    )
    async def set_scene_audio(
        project_id: str, scene_id: str, data: SceneAudioRequest
    ) -> Project:
        """Adjust a scene's gain and audio fades."""
        return guard_value(
            lambda: service.set_scene_audio(
                project_id, scene_id, data.volume, data.fade_in, data.fade_out
            )
        )

    @router.post(
        "/projects/{project_id}/timeline/scenes/{scene_id}/copy",
        response_model=SceneClipResponse,
    )
    async def copy_scene(project_id: str, scene_id: str) -> SceneClipResponse:
        """Copy a scene to a JSON clipboard the client can paste later."""
        return guard_value(
            lambda: SceneClipResponse(clip=service.copy_scene(project_id, scene_id))
        )

    @router.post("/projects/{project_id}/timeline/scenes/paste", response_model=Project)
    async def paste_scene(project_id: str, data: ScenePasteRequest) -> Project:
        """Paste a scene previously returned by the copy endpoint."""
        return guard_value(
            lambda: service.paste_scene(project_id, data.clip, data.after_scene_id)
        )

    @router.delete(
        "/projects/{project_id}/timeline/markers/{marker_id}", response_model=Project
    )
    async def remove_marker(project_id: str, marker_id: str) -> Project:
        return guard_value(
            lambda: service.remove_timeline_marker(project_id, marker_id)
        )

    @router.post(
        "/projects/{project_id}/video-project/ai-assist", response_model=Project
    )
    async def ai_assist(project_id: str, data: AiAssistRequest) -> Project:
        return guard(
            lambda: service.apply_ai_assist(
                project_id, fit=data.fit, beat=data.beat, bpm=data.bpm
            )
        )

    @router.get("/projects/{project_id}/video-project/scenes/{scene_id}/suggest")
    async def ai_suggest_scene(project_id: str, scene_id: str) -> dict[str, str]:
        try:
            return service.ai_suggest_scene(project_id, scene_id)
        except NotFoundError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except Exception as exc:
            raise HTTPException(status_code=409, detail=str(exc)) from exc

    @router.post(
        "/projects/{project_id}/video-project/scenes/{scene_id}/polish",
        response_model=Project,
    )
    async def polish_scene(project_id: str, scene_id: str) -> Project:
        return guard_value(lambda: service.polish_scene_text(project_id, scene_id))

    @router.post("/projects/{project_id}/video/upload", response_model=Project)
    async def upload_video(project_id: str, file: UploadFile = File(...)) -> Project:  # noqa: B008
        get_or_404(service, project_id)
        content = await file.read()
        try:
            return service.upload_video(
                project_id, file.filename or "video.webm", content
            )
        except Exception as exc:
            raise HTTPException(
                status_code=502, detail=f"Upload failed: {exc}"
            ) from exc

    @router.get("/projects/{project_id}/video")
    async def get_video(project_id: str, download: bool = False) -> FileResponse:
        get_or_404(service, project_id)
        path = service.video_path(project_id)
        if path is None:
            raise HTTPException(status_code=404, detail="No exported video yet")
        return FileResponse(
            path,
            media_type="video/mp4" if path.suffix == ".mp4" else "video/webm",
            filename=path.name if download else None,
        )

    @router.post("/projects/{project_id}/render", response_model=Project)
    def render_video(project_id: str, data: RenderRequest | None = None) -> Project:
        """Render the current timeline with the local ffmpeg adapter."""
        try:
            request = data or RenderRequest()
            return guard_value(
                lambda: service.render_video(
                    project_id, request.export_format, request.audio_ref
                )
            )
        except RenderError as exc:
            raise HTTPException(status_code=503, detail=str(exc)) from exc

    @router.post("/projects/{project_id}/voiceover/generate", response_model=Project)
    async def generate_voiceover(project_id: str) -> Project:
        return guard(lambda: service.generate_voiceover(project_id))

    @router.get("/projects/{project_id}/voiceover/{scene_id}")
    async def get_voiceover(project_id: str, scene_id: str) -> FileResponse:
        get_or_404(service, project_id)
        path = service.voiceover_path(project_id, scene_id)
        if path is None:
            raise HTTPException(status_code=404, detail="Voiceover clip not found")
        return FileResponse(path, media_type="audio/mpeg")

    @router.post(
        "/projects/{project_id}/timeline/extract",
        response_model=StructuredTimeline,
    )
    def extract_project_timeline(project_id: str) -> StructuredTimeline:
        """Extract a structured event timeline with turning points & casualties."""
        return guard_value(lambda: service.extract_timeline(project_id))

    return router

File: frontend/src/lib/api/timeline.ts
/**
 * Timeline endpoints: the video project, its scenes, markers and renders.
 *
 * Note on ``runTimelineCommand``: the request model lives in the backend's
 * ``models/qa.py`` (it is wired into the QA router), so its types sit in
 * ``@/types/qa`` even though the route is under ``/timeline``.
 */

import * as api from "@/lib/api/client";
import { Project } from "@/types/project";
import { StructuredTimeline } from "@/types/history";
import { TimelineCommandRequest, TimelineCommandResult } from "@/types/qa";
import {
  RenderPlan,
  TimelineMarker,
  TimelineReport,
  VideoProject,
} from "@/types/timeline";

/** ``GET /projects/{id}/video-project`` — the scenes, without the whole project. */
export function getVideoProject(projectId: string): Promise<VideoProject> {
  return api.apiFetch<VideoProject>(`/projects/${projectId}/video-project`);
}

/** ``PUT /projects/{id}/video-project`` — save the whole scene collection. */
export function saveVideoProject(
  projectId: string,
  project: VideoProject
): Promise<Project> {
  return api.apiFetch<Project>(`/projects/${projectId}/video-project`, {
    method: "PUT",
    body: { project },
  });
}

/** ``PUT /projects/{id}/timeline`` — save the timeline (an alias of the above). */
export function saveTimeline(
  projectId: string,
  timeline: VideoProject
): Promise<Project> {
  return api.apiFetch<Project>(`/projects/${projectId}/timeline`, {
    method: "PUT",
    body: timeline,
  });
}

/** ``GET /projects/{id}/timeline/report`` — measurements and validation findings. */
export function getTimelineReport(projectId: string): Promise<TimelineReport> {
  return api.apiFetch<TimelineReport>(`/projects/${projectId}/timeline/report`);
}

/** ``GET /projects/{id}/render-plan`` — the compiled, renderer-ready plan. */
export function getRenderPlan(projectId: string): Promise<RenderPlan> {
  return api.apiFetch<RenderPlan>(`/projects/${projectId}/render-plan`);
}

/** ``POST /projects/{id}/timeline/normalize`` — repair a timeline in place. */
export function normalizeTimeline(projectId: string): Promise<Project> {
  return api.apiFetch<Project>(`/projects/${projectId}/timeline/normalize`, {
    method: "POST",
  });
}

/** ``POST .../scenes/{scene_id}/split`` — razor at a fraction of the scene. */
export function splitScene(
  projectId: string,
  sceneId: string,
  at = 0.5
): Promise<Project> {
  return api.apiFetch<Project>(
    `/projects/${projectId}/timeline/scenes/${sceneId}/split`,
    { method: "POST", body: { at } }
  );
}

/** ``POST .../scenes/{scene_id}/merge`` — fold a scene into the previous one. */
export function mergeScene(projectId: string, sceneId: string): Promise<Project> {
  return api.apiFetch<Project>(
    `/projects/${projectId}/timeline/scenes/${sceneId}/merge`,
    { method: "POST" }
  );
}

/** ``POST .../scenes/{scene_id}/duplicate``. */
export function duplicateScene(
  projectId: string,
  sceneId: string
): Promise<Project> {
  return api.apiFetch<Project>(
    `/projects/${projectId}/timeline/scenes/${sceneId}/duplicate`,
    { method: "POST" }
  );
}

/** ``DELETE .../scenes/{scene_id}``. */
export function deleteScene(projectId: string, sceneId: string): Promise<Project> {
  return api.apiFetch<Project>(
    `/projects/${projectId}/timeline/scenes/${sceneId}`,
    { method: "DELETE" }
  );
}

/** ``POST .../scenes/{scene_id}/move`` — reorder a scene. */
export function moveScene(
  projectId: string,
  sceneId: string,
  toIndex: number
): Promise<Project> {
  return api.apiFetch<Project>(
    `/projects/${projectId}/timeline/scenes/${sceneId}/move`,
    { method: "POST", body: { to_index: toIndex } }
  );
}

/** ``POST .../timeline/scenes/bulk`` — one look applied to many scenes. */
export function bulkUpdateScenes(
  projectId: string,
  sceneIds: string[],
  patch: Record<string, unknown>
): Promise<Project> {
  return api.apiFetch<Project>(`/projects/${projectId}/timeline/scenes/bulk`, {
    method: "POST",
    body: { scene_ids: sceneIds, patch },
  });
}

/** ``POST .../timeline/markers``. */
export function addMarker(
  projectId: string,
  marker: Omit<TimelineMarker, "id">
): Promise<Project> {
  return api.apiFetch<Project>(`/projects/${projectId}/timeline/markers`, {
    method: "POST",
    body: marker,
  });
}

/** ``DELETE .../timeline/markers/{marker_id}``. */
export function removeMarker(
  projectId: string,
  markerId: string
): Promise<Project> {
  return api.apiFetch<Project>(
    `/projects/${projectId}/timeline/markers/${markerId}`,
    { method: "DELETE" }
  );
}

/** ``POST .../scenes/{scene_id}/speed``. */
export function setSceneSpeed(
  projectId: string,
  sceneId: string,
  speed: number
): Promise<Project> {
  return api.apiFetch<Project>(
    `/projects/${projectId}/timeline/scenes/${sceneId}/speed`,
    { method: "POST", body: { speed } }
  );
}

/** ``POST .../scenes/{scene_id}/reverse``. */
export function reverseScene(
  projectId: string,
  sceneId: string,
  reverse = true
): Promise<Project> {
  return api.apiFetch<Project>(
    `/projects/${projectId}/timeline/scenes/${sceneId}/reverse`,
    { method: "POST", body: { reverse } }
  );
}

/** ``POST .../scenes/{scene_id}/trim`` — clip in/out points on the source media. */
export function trimScene(
  projectId: string,
  sceneId: string,
  trim: { trim_start?: number | null; trim_end?: number | null }
): Promise<Project> {
  return api.apiFetch<Project>(
    `/projects/${projectId}/timeline/scenes/${sceneId}/trim`,
    { method: "POST", body: trim }
  );
}

/** ``POST .../scenes/{scene_id}/audio`` — narration gain and ramps. */
export function setSceneAudio(
  projectId: string,
  sceneId: string,
  audio: {
    volume?: number | null;
    fade_in?: number | null;
    fade_out?: number | null;
  }
): Promise<Project> {
  return api.apiFetch<Project>(
    `/projects/${projectId}/timeline/scenes/${sceneId}/audio`,
    { method: "POST", body: audio }
  );
}

/** ``POST .../scenes/{scene_id}/copy`` — snapshot a scene as a clip. */
export function copyScene(
  projectId: string,
  sceneId: string
): Promise<{ clip: Record<string, unknown> }> {
  return api.apiFetch<{ clip: Record<string, unknown> }>(
    `/projects/${projectId}/timeline/scenes/${sceneId}/copy`,
    { method: "POST" }
  );
}

/** ``POST .../timeline/scenes/paste``. */
export function pasteScene(
  projectId: string,
  clip: Record<string, unknown>,
  afterSceneId?: string
): Promise<Project> {
  return api.apiFetch<Project>(`/projects/${projectId}/timeline/scenes/paste`, {
    method: "POST",
    body: { clip, after_scene_id: afterSceneId ?? null },
  });
}

/** ``POST /projects/{id}/video-project/ai-assist`` — fit, beat and re-time. */
export function aiAssist(
  projectId: string,
  payload: { fit?: boolean; beat?: boolean; bpm?: number } = {}
): Promise<Project> {
  return api.apiFetch<Project>(`/projects/${projectId}/video-project/ai-assist`, {
    method: "POST",
    body: payload,
  });
}

/** ``GET .../scenes/{scene_id}/suggest`` — an AI suggestion for one scene. */
export function suggestScene(
  projectId: string,
  sceneId: string
): Promise<Record<string, string>> {
  return api.apiFetch<Record<string, string>>(
    `/projects/${projectId}/video-project/scenes/${sceneId}/suggest`
  );
}

/** ``POST .../scenes/{scene_id}/polish`` — rewrite one scene's text. */
export function polishScene(projectId: string, sceneId: string): Promise<Project> {
  return api.apiFetch<Project>(
    `/projects/${projectId}/video-project/scenes/${sceneId}/polish`,
    { method: "POST" }
  );
}

/** ``POST /projects/{id}/video/upload`` — attach an exported file. */
export function uploadVideo(projectId: string, file: File): Promise<Project> {
  const form = new FormData();
  form.append("file", file);
  return api.apiFetch<Project>(`/projects/${projectId}/video/upload`, {
    method: "POST",
    body: form,
  });
}

/** ``POST /timeline/command`` — apply a natural-language edit instruction. */
export function runTimelineCommand(
  payload: TimelineCommandRequest
): Promise<TimelineCommandResult> {
  return api.apiFetch<TimelineCommandResult>("/timeline/command", {
    method: "POST",
    body: payload,
  });
}

/** ``POST /projects/{id}/timeline/extract`` — the chronological event chain. */
export function extractTimeline(projectId: string): Promise<StructuredTimeline> {
  return api.apiFetch<StructuredTimeline>(
    `/projects/${projectId}/timeline/extract`,
    { method: "POST" }
  );
}

/** ``GET /projects/{id}/voiceover/{scene_id}`` — one narration clip's URL. */
export function voiceoverUrl(projectId: string, sceneId: string): string {
  const base = process.env.NEXT_PUBLIC_API_BASE_URL ?? "";
  return `${base}/projects/${projectId}/voiceover/${sceneId}`;
}